README.md
# [deprecated] Mapwize Android SDK using Webview rendering

This is the Mapwize Android SDK, version 2.x.

It allows you to display and interact with Mapwize venue maps.

## New SDK available

**This SDK remains functional and fully supported by Mapwize. However, a new SDK built on Mapbox GL is now available and all developers are invited to use it instead. Details are available at [docs.mapwize.io](https://docs.mapwize.io).**

## Requirements

Android API Level 15 (Android 4.0.3) or higher.

## Installation

Add in build.gradle

	repositories {
		maven {
			url "https://jitpack.io"
		}
	}

and 

	dependencies {
		compile 'com.github.Mapwize:mapwize-android-sdk:X.X.X'
	}

substitute x.x.x with ultimate version released

## Example application

An example application is provided in this repository, inside the example directory.

The menu lists most of the possible actions that you can do with the map.

## Documentation

### Setup api key

In order to use the map and the api manager, you have to intialize the MWZAccountManager with your api key. It can be obtained from the Mapwize Studio interface. If you don't have any, contact us at support@mapwize.io.

```java
MWZAccountManager.start(this, "YOUR API KEY");
```

If you don't want to use the Api manager and only the map, you can pass the api key in the xml layout as shown in the next part.

### Display the map
To display the Mapwize map, add it in your layout

```xml
<io.mapwize.mapwize.MWZMapView
android:id="@+id/mwzview"
android:layout_width="match_parent"
android:layout_height="match_parent"
app:center_latitude="51.508653"
app:center_longitude="-0.124975"
app:zoom="15"
app:floor="1.0"
/> 
```

The following custom attributes can be use :

```xml
<!-- Access -->
<attr name="accesskey" format="string"/>

<!-- Language -->
<attr name="language" format="string"/>

<!-- Controller display -->
<attr name="showUserPositionControl" format="boolean"/>
<attr name="displayFloorControl" format="boolean"/>
<attr name="mainColor" format="string"/>

<!-- Location -->
<attr name="isLocationEnabled" format="boolean"/>
<attr name="isBeaconsEnabled" format="boolean"/>

<!-- Coordinates -->
<attr name="center_latitude" format="float"/>
<attr name="center_longitude" format="float"/>
<attr name="zoom" format="integer"/>
<attr name="floor" format="float"/>
<attr name="minZoom" format="integer"/>

<!-- Bounds -->
<attr name="maxBounds_latitudeMin" format="float"/>
<attr name="maxBounds_latitudeMax" format="float"/>
<attr name="maxBounds_longitudeMin" format="float"/>
<attr name="maxBounds_longitudeMax" format="float"/>
<attr name="bounds_latitudeMin" format="float"/>
<attr name="bounds_latitudeMax" format="float"/>
<attr name="bounds_longitudeMin" format="float"/>
<attr name="bounds_longitudeMax" format="float"/>

<!-- Custom marker -->
<attr name="iconUrl" format="string"/>
<attr name="x_iconSize" format="integer"/>
<attr name="y_iconSize" format="integer"/>
<attr name="x_iconAnchor" format="integer"/>
<attr name="y_iconAnchor" format="integer"/>
```

### Life cycle
To avoid any memory leak, and in particular if you are using the iBeacon positioning solution within the Mapwize SDK, 
you need to call onDestroy on the mapview when its activity is destroyed. 

```java
@Override
protected void onDestroy() {
    super.onDestroy();
    mapview.onDestroy();
}
```

### Map options
Options are defined using the class MWZMapOptions. The following options are available:

- apiKey : [String] must be provided for the map to load, if it was not previously set in the MWZAccountManager. It can be obtained from the Mapwize Studio interface. If you don't have any, contact us at support@mapwize.io.
- bounds : [MWZLatLonBounds] region that will be display when map is loading
- maxBounds : [MWZLatLonBounds] region users are allowed to navigate in (default: entire world).
- center: [MWZLatLon] coordiantes of the center of the map at start-up (default: 0,0).
- zoom : [int] integer between 0 and 21 (default 0).
- minZoom: [int] optional minimum zoom allowed by the map, usefull to limit the visible area.
- floor : [double] double representing the desired floor of the building (default 0.0).
- isLocationEnabled : [boolean] boolean defining if the GPS should be started and the user position displayed (default: true).
- isBeaconsEnabled : [BOOL] boolean defining if the iBeacon scanner should be turned on (default: false).
- accessKey: [String] optional accessKey to be used during map load to be sure that access is granted to desired buildings at first map display.
- language: [String] optional preferred language for the map. Used to display all venues supporting that language.
- mainColor: [String] optional mainColor will be use to replace mapwize color in UI component as floor controller or direction.


### Moving the map
Once the map loaded, you can use the following functions on the map instance:

```java
public void fitBounds(MWZBounds bounds) 
public void centerOnCoordinates()
public void centerOnCoordinatesWithFloor()
public void setFloor(Double floor)
public void setZoom(Integer zoom)
public void centerOnVenue(MWZVenue venue)
public void centerOnVenue(String id)
public void centerOnPlace(MWZPlace place)
public void centerOnPlace(String id)
public void centerOnUser(Integer zoom)
```

### Getting map state

```java
public Integer getZoom()
public MWZCoordinate getCenter()
public Double getFloor()
public Double[] getFloors()
```

### User position

You can get/set the user position using the following methods. For a complete guide on the user position measurement principle, please refer to the [Mapwize.js documentation](https://github.com/Mapwize/mapwize.js-dist/blob/master/doc/doc.md).

```java
public MWZCoordinate getUserPosition()
public void newUserPositionMeasurement(MWZMeasurement measurement)
public void removeUserPosition()
public void setUserPosition(MWZUserPosition userPosition)
public void unlockUserPosition()
```
The followUserMode defines if the map should move when the user is moving.

```java
public boolean getFollowUserMode()
public void setFollowUserMode(boolean follow)
```

If you set isLocationEnabled=false in the map options, you can control the location using the functions
```java
public void startLocation(boolean useBeacon)
public void stopLocation()
```
### User heading

When a compass is available, it can be interesting to display the direction the user is looking at. To do so, the method setUserHeading can be used, giving it an angle in degree. To remove the display of the compass, simply set the angle to null.
```java
public void setUserHeading(Double heading)
```	

In the example app, you will find an example of code to listen to compass changes and display it on the map.

### Using Mapwize URLs

You can load Mapwize URLs using the following command. For a complete documentation please refer to the [Mapwize URL Scheme](https://github.com/Mapwize/mapwize-url-scheme).
```java
public void loadURL(String url)
```
### Adding markers

You can add markers on top of the map. A marker with floor null will be displayed on all floors

```java
public void addMarker(MWZCoordinate coordinate)
@Deprecated
public void addMarker(Double latitude, Double longitude, Double floor)
public void addMarker(String placeId)
public void removeMarkers()
```

You can change the default mapwize marker icon by passing options to the map. See map options.

### Controlling places display

You can promote places to increase their display priority.

```java
public void setPromotedPlaces(List<MWZPlace> places)
public void setPromotedPlacesWithIds(List<String> placeIds)
public void addPromotedPlace(MWZPlace place)
public void addPromotedPlaceWithId(String placeId)
public void addPromotedPlaces(List<MWZPlace> places)
public void addPromotedPlacesWithIds(List<String> placeIds)
public void removePromotedPlace(MWZPlace place)
public void removePromotedPlaceWithId(String placeId)
```

You can ignore place to make them not visible

```java
public void addIgnoredPlace(MWZPlace place)
public void addIgnoredPlaceWithId(String identifier)
public void removeIgnoredPlace(MWZPlace place)
public void removeIgnoredPlaceWithId(String identifier)
public void setIgnoredPlaces(List<MWZPlace> places)
public void setIgnoredPlacesWithIds(List<String> placeIds)
```

You can add places from your own data using this method. In order to make it works, you have to create a Mapwize formatted place object from your data and pass all the place as argument
```java
public void addExternalPlaces(List<MWZPlace> places)
```

The minimum attribute that should be specified to work propertly are the following :

MWZPlace.geometry
MWZPlace.venueId
MWZPlace.floor
MWZPlace.translations

In order to use promote and ignore methods, you have to specify an uniq id (MWZPlace.identifier)

### Universes

You can change the universe of a venue with the following methods.

```java
public void setUniverseForVenue(String universeId, MWZVenue venue)
public void setUniverseForVenue(MWZUniverse universe, MWZVenue venue)
```



### Showing directions

To start and stop directions you can you the following methods.

```java
public void startDirections(MWZDirection direction)
public void stopDirections()
```

The direction will be displayed and the event didStartDirections will be fired once the direction is loaded.
See the ApiManager part of this documentation to learn more about getting directions.

### Listening to events

You can listen for events emitted by the map using the MWZMapViewListener. To do so, set the listener using

```java
map.setListener(this);
```

then implement the following methods

```java
void onMapLoad();
void onMapLoaded();
void onReceivedError(String error);
void onZoomEnd(Integer zoom);
void onClick(MWZCoordinate latlon);
void onContextMenu(MWZCoordinate latlon);
void onFloorChange(Double floor);
void onFloorsChange(Double[] floors);
void onPlaceClick(MWZPlace place);
void onVenueClick(MWZVenue venue);
void onMarkerClick(MWZCoordinate position);
void onMoveEnd(MWZCoordinate latlon);
void onUserPositionChange(MWZUserPosition userPosition);
void onFollowUserModeChange(boolean followUserMode);
void onDirectionsStart(String info);
void onDirectionsStop(String info);
void onMonitoredUuidsChange(String[] uuids);
void onMissingPermission(String accessFineLocation);
void onJavascriptConsoleCallback(ConsoleMessage message);
```

onReveivedError will be triggered in case of unavilable internet connection or in case one of the interaction with the map failed.

### Editing place style

The display style of places can be changed dynamically within the SDK. To do so, you can use

```java
public void setStyle(String placeId,MWZStyle style)
```

### Setting top and bottom margins

It often happens that part of the map is hidden by banners or controls on the top or on the bottom. For example, if you display a banner to show the details of the place you just clicked on, it's better to display the banner on top of the map than having to resize the map.

However, you want to make sure that the Mapwize controls are always visible, like the followUserMode button and the floor selector. Also, that if you make a fitBounds, the area will be completely in the visible part of the map.

For this purpose, you can set a top and a bottom margin on the map. We garantee that nothing important will be displayed in those margin areas.

To set the margins in pixel:

```java
public void setBottomMargin(Integer margin)
public void setTopMargin(Integer margin)
```

### Refreshing the cache

To prevent too many network requests while browsing the map, the SDK keeps a cache of some data it already downloaded.
The Time To Live of the cache is 5 minutes.

If you want to force the map to refresh the cache and update itself, you can call the refresh method anytime. 
This method should be call each time a new access is granted by MWZApi.getAccess("mynewaccesskey")

```java
public void refresh()
```

## API Documentation

Mapwize objects and data can be retrieved without instantiating the map thanks to the MWZApi instance.

Before the ApiManager can be used, the api key need to be provided to the MWZAccountManager:
```java
MWZAccountManager.start(this, "YOUR API KEY");
```

### Api callback interface 

All api requests use the same callback interface. The type of the returned object depends on the request.
onFailure is called if an error occured
onSuccess is called in others cases

```java
public interface MWZCallback<T> {
void onSuccess(T object);
void onFailure(Throwable t);
}
```

### Accessing private venues

To access a private venue, enter the related accessKey using:
```java
public static void getAccess(String access, final MWZCallback<Map<String, Object>[]> callback)
```
Access keys are managed in the Mapwize Studio interface and are provided by venue managers.

Getting an access using api give the access in the map at the same time. If the map is already loaded when the call is done, you have to refresh the map.

### Getting venues

Venues have properties name and alias that allow to identify them in a more human-readable manner than the ID. Venue names and alias are unique throughout the Mapwize platform.

```java
public static void getVenues(Map<String,String> options, final MWZCallback<List<MWZVenue>> callback)
public static void getVenuesWithOrganizationId(@NonNull String organizationId, @NonNull final MWZCallback<List<MWZVenue>> callback)
public static void getVenueWithId(String id, final MWZCallback<MWZVenue> callback)
public static void getVenueWithName(String name, final MWZCallback<MWZVenue> callback)
public static void getVenueWithAlias(String alias, final MWZCallback<MWZVenue> callback)
```

### Getting places

Place have id, name and alias like venues. ID is unique throughout the mapwize platform but name and alias are unique throughout each venues

```java
public static void getPlaces(Map<String,String> options, final MWZCallback<List<MWZPlace>> callback)
public static void getPlaceWithId(String id, final MWZCallback<MWZPlace> callback)
public static void getPlaceWithName(String name, MWZVenue venue, final MWZCallback<MWZPlace> callback) 
public static void getPlaceWithAlias(String alias, MWZVenue venue, final MWZCallback<MWZPlace> callback)
public static void getPlacesForVenue(MWZVenue venue, final MWZCallback<List<MWZPlace>> callback)
```

### Getting placeLists

Placelist have the same behavior than places. You can use the following methods.

```java
public static void getPlacelists(Map<String,String> options, final MWZCallback<List<MWZPlaceList>> callback)
public static void getPlaceListWithId(String id, final MWZCallback<MWZPlaceList> callback)
public static void getPlaceListWithName(String name, MWZVenue venue, final MWZCallback<MWZPlaceList> callback)
public static void getPlaceListWithAlias(String alias, MWZVenue venue, final MWZCallback<MWZPlaceList> callback)
public static void getPlaceListsForVenue(MWZVenue venue, final MWZCallback<List<MWZPlaceList>> callback)
```
Placelist is a group of places so you may want to retrieve all the place object contained in th place list.

```java
public static void getPlacesForPlaceList(MWZPlaceList placeList, final MWZCallback<List<MWZPlace>> callback)
```

### Getting directions

To request a direction from one place to another, possibly with intermediate waypoints, the following method can be used:

```java
public static void getDirection(MWZDirectionPoint from, MWZDirectionPoint to, List<MWZDirectionPoint> waypoints, MWZDirectionOptions options, final MWZCallback<MWZDirection> callback)
public static void getDirection(MWZDirectionPoint from, List<MWZDirectionPoint> to, List<MWZDirectionPoint> waypoints, MWZDirectionOptions options, final MWZCallback<MWZDirection> callback)
```

- from (MWZPlace, MWZCoordinate,MWZCoordinateInVenue, MWZUserPosition) : The starting point of the direction
- to (MWZPlace, MWZPlaceList, MWZCoordinate,MWZCoordinateInVenue, MWZUserPosition) : Can be a MWZDirectionPoint or a List<MWZDirectionPoint>. In the first case, the to object will be the destination point. In the second case, the nearest point of the list will be use as destination point.
- waypoint (MWZPlace, MWZCoordinate,MWZCoordinateInVenue, MWZUserPosition) : A list of intermediate point.

Direction options :
- isAccessible (boolean) :  Boolean defining if the direction should avoid inaccessible way for someone with reduced mobility.
- waypointOptimize (boolean) : Boolean defining if the direction engine should optimize waypoint order to return the shortest way. 


### Search engine

MWZVenue, MWZPlace and MWZPlaceList are searchable object. They can be retrieve by the search method.

```java
public static void search(MWZSearchParams searchParams, final MWZCallback<List<MWZSearchableObject>> callback)
```

The search parameters are built as follow

```java
String query;           // The query string used for the search
String venueId;         // If specify, restrict the search to this venue
String organizationId;  // If specify, restrict the search to this organization
String universeId;      // If specify, restrict the search to this universe
```

## Proguard configuration

If you are using Proguard, add the following lines in yours proguard rules

-keepclassmembers class io.mapwize.mapwize.* {*;}
-keepattributes *Annotation*,EnclosingMethod,Signature
-keepnames class com.fasterxml.jackson.** { *; }
-dontwarn com.fasterxml.jackson.databind.**

## Contact

If you need any help, please contact us at contact@mapwize.io

## License

Mapwize Android SDK is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# CHANGELOG

## Version 2.3.8

- Adding possibility to set the server url

## Version 2.3.7

- Fixed google chrome update issue

## Version 2.3.6

- Fixed followUserMode change on floorChange from sdk
- Added method startDirection(Direction direction, boolean preventFitbounds)

## Version 2.3.5

- Api key must be passed in MWZAccountManager. Api key in xml does not work anymore
- Added event onVenueEnter and onVenueExit

## Version 2.3.4

- Fixed mainColor affects userPosition

## Version 2.3.3

- Floor, zoom and center map options now accessible using getters on the map after init
- Using Integer in MWZMapOptions instead of int
- Fixed bug with click on external place
- Limited required attributes for external place to : venueId, translations, floor and geometry
- Removed highlight on place click

## Version 2.3.2

- Fixed bug with options

## Version 2.3.1

- Fixed bug with bounds map option
- Fixed bug with zoom not set properly when passed as map option

## Version 2.3.0

- Added mainColor options
- Added support for external places
- Added option displayFloorControl
- Addes markerDisplay options

## Version 2.2.0

- Deprecated methods using latitude, longitude and floor to be replaced by methods using MWZCoordinate or MWZUserPosition
- Added a set of methods to promote places (make them displayed with highest priority)
- Added a set of methods to ignore places (hide them)
- Added onDestroy method on MWZMapView that has to be called on activity destroy when iBeacon location is used

## Version 2.1.4

- Added documentation for setUniverseForVenue methods
- Added map option to hide user position button
- Fixed bug with translation subTitle

## Version 2.1.3

- Fixed bug with the search menu in the demo app
- Added API call to retrieve universes for an organization, MWZMapView.setUniverseForVenue can now take a MWZUniverse 
- Deprecated onMapLoad method, even if still fired for now; Use onMapLoaded instead 
- Added getBounds method for geometries

## Version 2.1.2

- Added get methods to MWZSearchable
- Added search bar example in demo app

## Version 2.1.1

- Added API methods to get venues, places and placeLists
- Added geometry and other parameters to MWZVenue

## Version 2.1.0

- showdirection is deprecated. Use MWZApi.getDirections and MWZMapView.startDirections methods instead.
- MWZLatLon and MWZLatLonBounds are replaced by MWZCoordinate and MWZBounds.
- MWZPosition has been removed. MWZPlace, MWZPlaceList and MWZCoordinate now implement the MWZDirectionPoint interface.
- Introduced MWZApi, a wrapper to easily retrieve Mapwize data. All API methods that were previously linked to the map have been moved to MWZApi.
- Introduced MWZAccountManager to share API Key between MWZApi and MWZMapView
- The MWZPlace object now includes the geometry
- MWZPlaceStyle has been renamed in MWZStyle

## Version 1.7.2

- Fixed bug related to measurement parsing and directions
- Added proguard rules

## Version 1.7.1

- Improved location service. Requires bluetooth et bluetooth_admin permissions to use iBeacons.
- Introduced startLocation and stopLocation methods in MWZMapView.
- onMapLoad only then the map is fully loaded.
- Added minZoom and maxBounds map options.

## Version 1.6.0

- Added completionHandlers to all methods which can fail
- Added support for multilingual venues
- General improvements and bugfix

## Version 1.5.0

- Added locationEnabled and beaconsEnables in MapOptions
- Added placeList support with api methods and directions to a list

## Version 1.4.1

- Added method to stop showing directions.
- Added method to set user position with accuracy.
- Added methods to set top and bottom margins.
- Added the possibility to display user heading. See code in example app.
- Added methods to query venues and places on id, name and alias.
- Added caching to limit network traffic and resist to loss of connection. Call refresh on the map to force the update of the data.


example/src/main/java/io/mapwize/example/MainActivity.java
package io.mapwize.example;

import android.Manifest;
import android.content.pm.PackageManager;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.LayoutInflaterCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.arlib.floatingsearchview.FloatingSearchView;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.mapwize.mapwize.*;

public class MainActivity extends AppCompatActivity implements MWZMapViewListener, SensorEventListener, SearchListener, FloatingSearchView.OnMenuItemClickListener{

    MWZMapView mapview;
    private SensorManager mSensorManager;
    private Sensor mCompass;
    private SearchFragment mSearchFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            if (!ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 5);
            }
        }
        mapview = (MWZMapView) this.findViewById(R.id.mwzview);
        mapview.setListener(this);
        mSearchFragment = (SearchFragment)getSupportFragmentManager().findFragmentById(R.id.search_fragment);
        mSearchFragment.setSearchListener(this);
        mSearchFragment.setMenuListener(this);

    }

    public void showToast(String text) {
        Toast toast = Toast.makeText(this, text, Toast.LENGTH_SHORT);
        toast.show();
    }

    public void setPreferredLanguageFR () {
        this.mapview.setPreferredLanguage("fr");
    }

    public void setPreferredLanguageEN () {
        this.mapview.setPreferredLanguage("en");
    }

    public void setZoom(){
        this.mapview.setZoom(12);
    }

    public void centerOnCoordinates() {
        MWZCoordinate coordinate = new MWZCoordinate(49.74252973220731, 4.599119424819946);
        this.mapview.centerOnCoordinates(coordinate, 18);
    }

    public void centerOnCoordinatesWithFloor() {
        MWZCoordinate coordinate = new MWZCoordinate(49.74252973220731, 4.599119424819946, 2d);
        this.mapview.centerOnCoordinates(coordinate, 18);
    }

    public void setFloor() {
        this.mapview.setFloor(2);
    }

    public void centerOnPlace() {
        this.mapview.centerOnPlace("56c3426202275a0b00fb00b9");
    }

    public void centerOnUser() {
        this.mapview.centerOnUser(19);
    }

    public void loadUrl() {
        this.mapview.loadURL("http://mwz.io/aaa", new LoadURLCallbackInterface() {
            @Override
            public void onResponse(Error error) {
                if (error != null) {
                    showToast("LoadUrlError");
                }
            }
        });
    }

    public void addMarker() {
        this.mapview.addMarker(49.74278626088478, 4.598293304443359, null);
    }

    public void removeMarkers() {
        this.mapview.removeMarkers();
    }

    public void setFollowUserModeON() {
        this.mapview.setFollowUserMode(true);
    }

    public void setFollowUserModeOFF() {
        this.mapview.setFollowUserMode(false);
    }

    public void setUserPosition() {
        MWZUserPosition userPosition = new MWZUserPosition();
        userPosition.setLatitude(49.74278626088478);
        userPosition.setLongitude(4.5982933044);
        userPosition.setAccuracy(50);
        this.mapview.setUserPosition(userPosition);
    }

    public void removeUserPosition() {
        this.mapview.removeUserPosition();
    }

    public void unlockUserPosition() {
        this.mapview.unlockUserPosition();
    }

    public void showDirections() {
        MWZCoordinate coord = new MWZUserPosition(49.74271259569392, 4.599329978227616, 0d, 15);
        MWZPlace from = new MWZPlace();
        from.setIdentifier("56c3426202275a0b00fb00b9");
        MWZPlace to = new MWZPlace();
        to.setIdentifier("56c3504102275a0b00fb00fa");
        MWZPlace wp1 = new MWZPlace();
        wp1.setIdentifier("56c34fc402275a0b00fb00f6");
        MWZPlace wp2 = new MWZPlace();
        wp2.setIdentifier("56c344a402275a0b00fb00bf");
        List<MWZDirectionPoint> arr = new ArrayList<>();
        arr.add(wp1);
        arr.add(wp2);
        MWZDirectionOptions options = new MWZDirectionOptions();
        options.setAccessible(true);
        MWZApi.getDirection(coord, to, arr, options, new MWZCallback<MWZDirection>() {
            @Override
            public void onSuccess(MWZDirection object) {
                mapview.startDirections(object);
            }

            @Override
            public void onFailure(Throwable t) {
                Log.i("Failure ", " " + t);
            }
        });

    }

    public void stopDirections() {
        this.mapview.stopDirections();
    }

    public void setStyle() {
        MWZStyle style = new MWZStyle();
        style.setFillColor("#FF0000");
        style.setLabelBackgroundColor("#0000FF");
        style.setStrokeColor("#AAAAAA");
        style.setMarkerUrl("https://cdn4.iconfinder.com/data/icons/medical-soft-1/512/map_marker_pin_pointer_navigation_location_point_position-128.png");
        this.mapview.setStyle("56c3426202275a0b00fb00b9", style);
    }

    public void setBottomMargin() {
        this.mapview.setBottomMargin(300);
    }

    public void setTopMargin() {
        this.mapview.setTopMargin(60);
    }

    public void resetMargin() {
        this.mapview.setBottomMargin(0);
        this.mapview.setTopMargin(0);
    }

    @SuppressWarnings("deprecation")
    public void setUserHeading() {
        mSensorManager = (SensorManager)getSystemService(SENSOR_SERVICE);
        mCompass = mSensorManager.getDefaultSensor(Sensor.TYPE_ORIENTATION);
        mSensorManager.registerListener(this, mCompass, SensorManager.SENSOR_DELAY_GAME);
    }

    public void removeHeading() {
        mSensorManager.unregisterListener(this);
        this.mapview.setUserHeading(null);
    }

    public void refresh() {
        this.mapview.refresh();
    }

    public void getFloor() {
        showToast("Floor :" + this.mapview.getFloor());
    }

    public void getZoom() {
        showToast("Zoom :" + this.mapview.getZoom());
    }

    public void getUserPosition() {
        showToast("UserPosition :" + this.mapview.getUserPosition());
    }

    public void getCenter() {
        showToast("Center :" + this.mapview.getCenter());
    }

    public void startLocation() {
        this.mapview.startLocation(true);
    }

    public void stopLocation() {
        this.mapview.stopLocation();
    }


    @Override
    public void onVenueSuggestionClicked(MWZVenue venue) {
        MWZCoordinate center = venue.getMarker();
        mapview.centerOnCoordinates(center, 17);
    }

    @Override
    public void onPlaceSuggestionClicked(final MWZPlace place) {
        MWZCoordinate center = place.getMarker();
        mapview.centerOnCoordinates(center, 21);
    }

    @Override
    public void onMapLoad() {
        Log.i("onMapLoad", "loaded");
    }

    @Override
    public void onZoomEnd(Integer zoom) {
        Log.i("Zoomend", "" + zoom);
    }

    @Override
    public void onClick(MWZCoordinate latlon) {
        Log.i("OnClick", "" + latlon);
    }

    @Override
    public void onContextMenu(MWZCoordinate latlon) {
        Log.i("OnContextMenu", "" + latlon);
    }

    @Override
    public void onFloorChange(Double floor) {
        Log.i("OnFloorChange", "" + floor);
    }

    @Override
    public void onFloorsChange(Double[] floors) {
        Log.i("OnFloorsChange", "floors");
    }

    @Override
    public void onPlaceClick(MWZPlace place) {
        Log.i("OnPlaceClick", "" + place);
    }

    @Override
    public void onVenueClick(MWZVenue venue) {
        Log.i("OnVenueClick", "" + venue);
    }

    @Override
    public void onMarkerClick(MWZCoordinate position) {
        Log.i("OnMarkerClick", "" + position);
    }

    @Override
    public void onMoveEnd(MWZCoordinate latlon) {
        Log.i("OnMoveEnd", "" + latlon);
    }

    @Override
    public void onUserPositionChange(MWZUserPosition coordinate) {
        Log.i("OnUserPositionChange", "" + coordinate);
    }

    @Override
    public void onFollowUserModeChange(boolean followUserMode) {
        Log.i("OnFollowUserModeChange", "" + followUserMode);
    }

    @Override
    public void onDirectionsStart(String info) {
        Log.i("OnDirectionsStart", "" + info);
    }

    @Override
    public void onDirectionsStop(String info) {
        Log.i("OnDirectionsStop", "" + info);
    }

    @Override
    public void onMonitoredUuidsChange(String[] uuids) {
        Log.i("OnMonitoredUuidsChange", "" + uuids);
    }

    @Override
    public void onMissingPermission(String accessFineLocation) {
        Log.i("onMissingPermission", "" + accessFineLocation);
    }

    @Override
    public void onMapLoaded() {
        Log.i("onMapLoaded", "MapWize loaded !");
        //mapview.refresh();
    }

    @Override
    public void onVenueEnter(MWZVenue venue) {

        Log.i("onVenueEnter", "" + venue.toString());
    }

    @Override
    public void onVenueExit(MWZVenue venue) {
        Log.i("onVenueExit", "done");
    }


    @Override
    public void onReceivedError(String error) {
        Log.i("OnReceivedError", error);
    }

    @Override
    @SuppressWarnings("deprecation")
    public final void onSensorChanged(SensorEvent event) {
        switch(event.sensor.getType()){
            case Sensor.TYPE_ORIENTATION:
                this.mapview.setUserHeading((new Double(event.values[0])));
                break;
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }

    @Override
    public void onActionMenuItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_setPreferredLanguageFR:
                this.setPreferredLanguageFR();
                break;
            case R.id.action_setPreferredLanguageEN:
                this.setPreferredLanguageEN();
                break;
            case R.id.action_setZoom:
                this.setZoom();
                break;
            case R.id.action_centerOnCoordinates:
                this.centerOnCoordinates();
                break;
            case R.id.action_centerOnCoordinatesWithFloor:
                this.centerOnCoordinatesWithFloor();
                break;
            case R.id.action_setFloor:
                this.setFloor();
                break;
            case R.id.action_centerOnPlace:
                this.centerOnPlace();
                break;
            case R.id.action_centerOnUser:
                this.centerOnUser();
                break;
            case R.id.action_loadUrl:
                this.loadUrl();
                break;
            case R.id.action_addMarker:
                this.addMarker();
                break;
            case R.id.action_removeMarkers:
                this.removeMarkers();
                break;
            case R.id.action_setFollowUserModeON:
                this.setFollowUserModeON();
                break;
            case R.id.action_setFollowUserModeOFF:
                this.setFollowUserModeOFF();
                break;
            case R.id.action_setUserPosition:
                this.setUserPosition();
                break;
            case R.id.action_unlockUserPosition:
                this.unlockUserPosition();
                break;
            case R.id.action_showDirections:
                this.showDirections();
                break;
            case R.id.action_stopDirections:
                this.stopDirections();
                break;
            case R.id.action_setStyle:
                this.setStyle();
                break;
            case R.id.action_setBottomMargin:
                this.setBottomMargin();
                break;
            case R.id.action_setTopMargin:
                this.setTopMargin();
                break;
            case R.id.action_resetMargin:
                this.resetMargin();
                break;
            case R.id.action_setUserHeading:
                this.setUserHeading();
                break;
            case R.id.action_removeHeading:
                this.removeHeading();
                break;
            case R.id.action_getFloor:
                this.getFloor();
                break;
            case R.id.action_getZoom:
                this.getZoom();
                break;
            case R.id.action_getUserPosition:
                this.getUserPosition();
                break;
            case R.id.action_getCenter:
                this.getCenter();
                break;
            case R.id.action_removeUserPosition:
                this.removeUserPosition();
                break;
            case R.id.action_startLocation:
                this.startLocation();
                break;
            case R.id.action_stopLocation:
                this.stopLocation();
                break;
            case R.id.action_refresh:
                this.refresh();
                break;
            }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mapview.onDestroy();
    }
}
